sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,624 @@
+    }
+}
+
+gpio_result_t gpio_intr_en_lvl_low (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_LVL_LOW_EN0), 1, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_dis_rise (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_RISE_EN0), 0, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_dis_fall (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_FALL_EN0), 0, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_dis_lvl_high (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_LVL_HIGH_EN0), 0, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_dis_lvl_low (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_LVL_LOW_EN0), 0, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_dis_all (gpio_pin_number_t pin){
+    if (pin >= 0 && pin < MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->INTRPT_RISE_EN0),     0, 1, pin);
+        setBitfield(&(gpio_peri->INTRPT_FALL_EN0),     0, 1, pin);
+        setBitfield(&(gpio_peri->INTRPT_LVL_HIGH_EN0), 0, 1, pin);
+        setBitfield(&(gpio_peri->INTRPT_LVL_LOW_EN0),  0, 1, pin);
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+
+}
+
+gpio_result_t gpio_intr_en (gpio_pin_number_t pin, gpio_intr_type_t type)
+{
+    if (pin >= 0 && pin < MAX_PIN)
+    { 
+        gpio_intr_dis_all(pin);
+
+        switch(type)
+        {
+        case GpioIntrEdgeRising:
+            gpio_intr_en_rise(pin);
+            break;
+
+        case GpioIntrEdgeFalling:
+            gpio_intr_en_fall(pin);
+            break;
+
+        case GpioIntrLevelLow:
+            gpio_intr_en_lvl_low(pin);
+            break;
+
+        case GpioIntrLevelHigh:
+            gpio_intr_en_lvl_high(pin);
+            break;
+
+        case GpioIntrEdgeRisingFalling:
+            gpio_intr_en_rise(pin);
+            gpio_intr_en_fall(pin);
+            break;
+
+        case GpioIntrEdgeRisingLevelLow:
+            gpio_intr_en_rise(pin);
+            gpio_intr_en_lvl_low(pin);
+            break;
+
+        case GpioIntrEdgeFallingLevelHigh:
+            gpio_intr_en_fall(pin);
+            gpio_intr_en_lvl_high(pin);
+            break;
+
+        default:
+            return GpioError;
+        }
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_intr_check_stat_rise (gpio_pin_number_t pin, bool *is_pending)
```

Comment:
I suggest avoiding passing reference to bool. Instead, the return value could be used to codify that, at the end you are using the return value to check almost as that

Reply:
if there is a need to codify more gpio_result_t enums, do it

Reply:
Actually, I prefer it this way, one gpio_result for the sanity check. I will add to it if other situations should be considered. But the values should pass to a user give it back through pointers that are passed. what do you think?!

Reply:
you can let the pointer as a reference if you want, but you do not need the if else statement, you can use the direct output of the read operation to set the `is_pending`